File: android/src/main/kotlin/com/trackingsport/geoar/FlutterGeoArPlugin.kt
package com.trackingsport.geoar

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** FlutterGeoArPlugin */
class FlutterGeoArPlugin :
    FlutterPlugin,
    MethodCallHandler {
    // The MethodChannel that will the communication between Flutter and native Android
    //
    // This local reference serves to register the plugin with the Flutter Engine and unregister it
    // when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel
    private lateinit var sensorEventChannel: EventChannel
    private var context: Context? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_geo_ar")
        channel.setMethodCallHandler(this)
        
        // Registrar EventChannel para sensores con optimizaciones
        sensorEventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "flutter_geo_ar/sensors")
        sensorEventChannel.setStreamHandler(SensorEventStreamHandler(context!!))
    }

    override fun onMethodCall(
        call: MethodCall,
        result: Result
    ) {
        if (call.method == "getPlatformVersion") {
            result.success("Android ${android.os.Build.VERSION.RELEASE}")
        } else {
            result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        sensorEventChannel.setStreamHandler(null)
        context = null
    }
}

File: ios/Classes/FlutterGeoArPlugin.swift
import Flutter
import UIKit

public class FlutterGeoArPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "flutter_geo_ar", binaryMessenger: registrar.messenger())
    let instance = FlutterGeoArPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
    
    // Registrar EventChannel para sensores con optimizaciones
    let sensorEventChannel = FlutterEventChannel(name: "flutter_geo_ar/sensors", binaryMessenger: registrar.messenger())
    let sensorStreamHandler = SensorStreamHandler()
    sensorEventChannel.setStreamHandler(sensorStreamHandler)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getPlatformVersion":
      result("iOS " + UIDevice.current.systemVersion)
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}
